### PR TITLE
release-21.1: kvserver: stop transferring leases to replicas that may need snapshots

### LINKED ISSUE #61604: kvserver: lease transferred to uninitialized follower
**Describe the problem**

In https://github.com/cockroachdb/cockroach/issues/61541#issuecomment-792670508, we saw that a range with an uninitialized follower transferred the lease to that uninitialized follower, leading to the range becoming unavailable until the follower caught up. This indicates a hole in our safety mechanisms.

**To Reproduce**

Unclear, but the test in #61541 definitely achieved it.

**Expected behavior**

**Additional data / screenshots**

**Environment:**

https://github.com/cockroachdb/cockroach/commits/b703e663da8ededaee2e28fc39a24e3880ae54cf

**Additional context**

It looks like aggressive changes in range size can lead to temporary (but substantial) unavailability. 

gz#9425

### PR DESCRIPTION
**Pull request by aayushshah15**
Backport 1/1 commits from #69696.

/cc @cockroachdb/release

---

This commit disallows the `replicateQueue` from initiating lease
transfers to replicas that may be in need of a raft snapshot. Note that
the `StoreRebalancer` already has a stronger form of this check since it
disallows lease transfers to replicas that are lagging behind the raft
leader (which includes the set of replicas that need a snapshot).

In cases where the raft leader is not the leaseholder, we disallow the
replicateQueue from any sort of lease transfer until leaseholdership and
leadership are collocated. We rely on calls to
`maybeTransferRaftLeadershipToLeaseholderLocked()` (called on every raft
tick) to make sure that such periods of leadership / leaseholdership
misalignment are ephemeral and rare.

Alternative to https://github.com/cockroachdb/cockroach/pull/63507

Release justification: bug fix

Resolves #61604

Release note (bug fix): Fixes a bug that can cause prolonged
unavailability due to lease transfer to a replica that may be in need of
a raft snapshot.

### CODE REVIEW

**Review by erikgrinaker — APPROVED** [review]
LGTM, but maybe we should respect the baking period here since we haven't made any hard promises for 21.1?

**Comment by aayushshah15** [comment]
>LGTM, but maybe we should respect the baking period here since we haven't made any hard promises for 21.1?

I agree, I'll wait until late next week to merge.

**Review by nvb — APPROVED** [review]
:lgtm:

> Reviewed 3 of 3 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/69964#-:-MjBt7cL-SlEvvg3QWOu:bnfp4nl)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @aayushshah15)

**Pull request merged**
